### ISSUE TITLE
"Some required modules must be enabled" page re-wording/formatting

### ISSUE THREAD

**Issue by BWPanda** [backdrop/backdrop-issues#3176]
When you enable a module that has dependencies that are disabled, you get a page like the following:

> ### Some required modules must be enabled
> - You must enable the ImgAreaSelect, imagesLoaded modules to install Manual Crop.
>
> Would you like to continue with the above?

If there are a lot of dependencies, this list can be hard to read.

I'd like to suggest re-wording/formatting this page as follows:

> ### Some required modules must be enabled
> - You must enable the following modules to install Manual Crop:
>     - ImgAreaSelect
>     - imagesLoaded
>
> Would you like to continue with the above?

i.e. Move the list of dependencies from a comma-separated list in the sentence to a bullet list below it.

**Comment by BWPanda** [comment]
One benefit of this change is that you don't need to worry about the whole singular/plural thing, as my proposed wording works just as well for single dependencies:

> ### Some required modules must be enabled
> - You must enable the following modules to install Menu Admin per Menu:
>     - Menu
>
> Would you like to continue with the above?

**Comment by klonos** [comment]
How's this?:

<img width="537" alt="screen shot 2018-06-26 at 8 26 12 pm" src="https://user-images.githubusercontent.com/2423362/41906123-653de692-7980-11e8-8289-126bd6de4cad.png">

**Comment by BWPanda** [comment]
Nice! Good job :+1:

**Comment by klonos** [comment]
...I was not happy with the fact that the way this was laid out, the user's eye would have to "travel" quite a bit right/left if they wanted to quickly scan the page:

<img width="537" alt="screen shot 2018-06-26 at 8 26 12 pm" src="https://user-images.githubusercontent.com/2423362/41907562-95cc7748-7984-11e8-9db1-2e41f17d575f.png">

So I have reworded this a bit further in order to keep the module names in the phrases as close to the dependencies lists as possible:

<img width="666" alt="screen shot 2018-06-26 at 9 02 11 pm" src="https://user-images.githubusercontent.com/2423362/41909232-d676f57a-7989-11e8-92ae-bf9092445bd9.png">

Still not 100% happy with it, so I will work on another PR. In the meantime, please keep the feedback coming.

**Comment by klonos** [comment]
I was experimenting with a table instead of a list with nested items:

<img width="452" alt="screen shot 2018-06-26 at 10 22 15 pm" src="https://user-images.githubusercontent.com/2423362/41911246-7a039504-798f-11e8-9f11-ca2fa8835506.png">

I find it way much easier to be scanning the list of modules to be enabled this way. Let me know what you think.

**Comment by klonos** [comment]
...I have also tried a list of "`requirement` (required by `module`)", but the repetition of "(required by `module`)" was making things too wordy for modules with multiple requirements. The table seemed to be having the same effect with added benefits:

- all modules to be enabled are distinctly listed to the left, separated in their own column, so they are easy to scan quickly, plus the zebra striping helps.
- the "required by" modules are merged into a single cell for multiple requirements, so there's less "clutter"
- we can set the priority of the "required by" column so that it gets hidden in narrow screens

In general it seem tidier 😄

**Comment by BWPanda** [comment]
The following feedback is based on the screenshots provided (since the PR doesn't have the table feature committed yet):

- I like the table layout in general; it looks cleaner and is much better for lots of dependencies/modules, but I wonder if it'll be overkill for the 80% when you just have one module requiring just one other module...?
- If it's decided to go back to the bullet points layout instead of a table, maybe consider the following wording:

> - The *Content Translation* module requires the following modules to also be enabled:
>     - Locale
>     - Language

Or something along those lines.
- Back to the table, I like the merging of the 'Required by' column, but I'd just want to make sure those cells are vertically-aligned top, otherwise lots of dependencies will push the 'Required by' module down a bit (vertically centered), which looks weird IMO.
- I'm not sure about the wording of the column names... What about 'Required module(s)' instead of 'Additional modules...', to match the heading of the page?

**Comment by klonos** [comment]
> ...I wonder if it'll be overkill for the 80% when you just have one module requiring just one other module...?

Same goes for the bulleted list 😄 

> ...maybe consider the following wording

Done! Please test the updated PR.

> ...I'd just want to make sure those cells are vertically-aligned top, otherwise ... looks weird IMO.

What I was going for was this visual/metal effect:

![image](https://user-images.githubusercontent.com/2423362/41938105-769ebd12-79d5-11e8-8b3a-63ed07db7657.png)

...notice how the bracket groups the list of multiple items, and then the point of the bracket points to a single item? I feel that not having this single item vertically centered (i.e. aligned top/bottom) seems odd:

![image](https://user-images.githubusercontent.com/2423362/41938175-acf7c016-79d5-11e8-8a70-79c4fdecb68d.png)

If others feel that align-top is more readable, I am happy to change that. The current implementation is what comes default OOTB with core; no additional styling has been done. For example, another thing that perhaps needs "fixing", is that on wider screens the distance between the 2 columns is greater, which might make things appear "disjoined":

<img width="1180" alt="screen shot 2018-06-27 at 6 47 35 am" src="https://user-images.githubusercontent.com/2423362/41938327-1f6eb8a2-79d6-11e8-8de3-aeb3b37d7025.png">

...so we should perhaps consider reducing/limiting the width of the table (is we go with that solution that is).

> What about 'Required module(s)' instead of 'Additional modules...'

Done! See screenshot above 😉

**Comment by docwilmot** [comment]
This is not a design pattern that core or contrib uses for confirm forms anywhere else. See any function with the word "multiple" in [this list of confirm forms](https://api.backdropcms.org/api/backdrop/core%21modules%21system%21system.module/function/calls/confirm_form/1).  And I think it is definitely overkill. I prefer and recommend the item list as in @BWPanda's first post, with whatever wording is appropriate.

**Comment by klonos** [comment]
I have filed a separate PR for the table implementation:

- formatted as bulleted list: https://github.com/backdrop/backdrop/pull/2230
- formatted as a table: https://github.com/backdrop/backdrop/pull/2231

Feel free to test both PR sandboxes and let me know which one seems best. Thanks

PS: I am exploring another option, so might have a 3rd PR coming up soon(ish) 😄

**Comment by klonos** [comment]
...so here's the thing: The things that should be treated as "first class" here are the required modules. If we render this as a list of "parent" modules, with the to-be-enabled modules nested within them (as per @BWPanda's initial suggestion), then the focus goes to the wrong things IMO.

Without touching much of what we currently have, I have tried to make them bold:

<img width="704" alt="screen shot 2018-06-27 at 9 21 48 am" src="https://user-images.githubusercontent.com/2423362/41944549-a55f0e3e-79eb-11e8-84e4-93cd3952dc35.png">

There's too much text being repeated there, and rendering the requirements as a list is so much better than shoving all of them in a single line. So, I have tried rendering this as a list with the required modules to be enabled listed first (also reduced the amount of text to a very minimum):

<img width="473" alt="screen shot 2018-06-27 at 8 59 49 am" src="https://user-images.githubusercontent.com/2423362/41943877-aea0e088-79e8-11e8-9df9-330bde4f252a.png">

...but as you can see, the amount of text there still makes it hard to quickly scan which modules are about to be enabled (which is the problem we are trying to solve here). So I have then tried making them bold:

<img width="480" alt="screen shot 2018-06-27 at 8 58 47 am" src="https://user-images.githubusercontent.com/2423362/41943945-f2e2412e-79e8-11e8-8c2b-361358f226ee.png">

...the problem with this iteration is that the "required by" lines are still being repeated. Although the module name is rendered in italics, it does not help much. So I have then tried adding this line only to the last requirement (per module), and adding dots for the ones before it:

<img width="477" alt="screen shot 2018-06-27 at 8 57 41 am" src="https://user-images.githubusercontent.com/2423362/41944053-57ada45e-79e9-11e8-9e5a-5786d6a99e68.png">

Umm, still don't like it 😐 ...because the same thing would look so much better as a table.

**Comment by klonos** [comment]
...pinging @wesruv for some input.

**Comment by herbdool** [comment]
I like the table best. Easiest to read

**Comment by BWPanda** [comment]
@klonos Are you able to provide screenshots of the following?

- Table with 5+ dependencies with 'Required by' module vertically centred
- Table with 5+ dependencies with 'Required by' module vertically-aligned top
- Bullets with one dependency for one module
- Bullets with multiple dependencies for multiple modules
- Table with one dependency for one module
- Table with multiple dependencies for multiple modules

That might help us better compare how the UI options look.

Also, regarding my proposed 'Required module(s)' wording, I meant that to be the whole text, not to keep the 'to be enabled' part after it... 🙂

Thanks for all your work on this!

**Comment by BWPanda** [comment]
Also, from your third option, I like the second-last screenshot the best, but I don't think we need the words 'the' or 'module' (e.g. make it simply "Locale (required by Content Translation)" with appropriate bold and italics).

**Comment by klonos** [comment]
> ...regarding my proposed 'Required module(s)' wording, I meant that to be the whole text, not to keep the 'to be enabled' part after it... 🙂

Ooops 😅 ...fixed in the PR.

Screenshots...

**Comment by klonos** [comment]
Table with 5+ dependencies aligned middle/top:

<img width="947" alt="screen shot 2018-06-27 at 5 28 22 pm" src="https://user-images.githubusercontent.com/2423362/41959608-d1e4bf18-7a30-11e8-96d1-609d328ff31e.png">

**Comment by klonos** [comment]
Table with one dependency for one module:

<img width="519" alt="screen shot 2018-06-27 at 5 41 19 pm" src="https://user-images.githubusercontent.com/2423362/41959810-60a1e866-7a31-11e8-903b-78e7fe4a53e6.png">

...for multiple dependencies, please refer to previous comment/screenshot.

**Comment by klonos** [comment]
...bullets with 
<img width="450" alt="screen shot 2018-06-27 at 5 43 45 pm" src="https://user-images.githubusercontent.com/2423362/41960193-7cc13294-7a32-11e8-8c9a-eea0b6e6d2d2.png">
single module/requirement

**Comment by klonos** [comment]
...bullets with many requirements and dots:

<img width="455" alt="screen shot 2018-06-27 at 5 47 38 pm" src="https://user-images.githubusercontent.com/2423362/41960240-9eab6cc6-7a32-11e8-965a-90d5e84cc250.png">

**Comment by klonos** [comment]
...bullets with many requirements and repeating "required by ...":

<img width="498" alt="screen shot 2018-06-27 at 5 49 09 pm" src="https://user-images.githubusercontent.com/2423362/41960279-b67f5330-7a32-11e8-81c5-bdf5de8fef8d.png">

**Comment by BWPanda** [comment]
Thanks @klonos!

I don't think one module is going to have that many dependencies for the vertically-centered table layout to be an issue, so if the table is what others decide to go ahead with, I'm ok with the vertically-centered option.

I definitely don't like the "bullets with many requirements and dots" option. The others are fine though.

**Comment by klonos** [comment]
Thanks for the feedback @BWPanda. Anyone else please?

**Comment by yorkshire-pudding** [comment]
@klonos @BWPanda 
I prefer the table for when there are many and I don't think there is any problem with it if there's only 1.  It looks clearer all round.
Thanks
